File: Java/876. Middle of the Linked List.java
// 0ms, 100.00%; 36.8MB, 7.84%
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public static ListNode middleNode(ListNode head) {
        ListNode fast = head, slow = head;
        while (fast.next != null) {
            fast = fast.next;
            if (fast.next != null) {
                fast = fast.next;
            }
            slow = slow.next;
        }
        return slow;
    }
}

File: Python3/876. Middle of the Linked List.py
# 28ms, 83.03%; 12.6MB, 100.00%
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        fast = head
        while fast.next:
            fast = fast.next
            if fast.next: fast = fast.next
            head = head.next
        return head
